src/coreclr/jit/instr.cpp

```diff
@@ -2024,31 +2024,38 @@ instruction CodeGenInterface::ins_Store(var_types dstType, bool aligned /*=false
 // Return Value:
 //   the instruction to use
 //
-// Notes:
-//   The function currently does not expect float srcReg with integral dstType and will assert on such cases.
-//
 instruction CodeGenInterface::ins_StoreFromSrc(regNumber srcReg, var_types dstType, bool aligned /*=false*/)
 {
+    assert(srcReg != REG_NA);
+
     bool dstIsFloatType = isFloatRegType(dstType);
     bool srcIsFloatReg  = genIsValidFloatReg(srcReg);
     if (srcIsFloatReg == dstIsFloatType)
     {
         return ins_Store(dstType, aligned);
     }
-
-    assert(!srcIsFloatReg && dstIsFloatType && "not expecting an integer type passed in a float reg");
-    assert(!varTypeIsSmall(dstType) && "not expecting small float types");
-
-    instruction ins = INS_invalid;
-#if defined(TARGET_XARCH)
-    ins = INS_mov;
-#elif defined(TARGET_ARMARCH)
-    ins     = INS_str;
-#else
-    NYI("ins_Store");
-#endif
-    assert(ins != INS_invalid);
-    return ins;
+    else
+    {
+        // We know that we are writing to memory, so we can treat destination as any type,
+        // so we will treat it as the source type.
+        var_types srcTypeForStore = TYP_UNDEF;
+        unsigned  dstSize         = genTypeSize(dstType);
+        switch (dstSize)
+        {
+            case 4:
+                srcTypeForStore = srcIsFloatReg ? TYP_FLOAT : TYP_INT;
+                break;
+#if defined(TARGET_64BIT)
+            case 8:
+                srcTypeForStore = srcIsFloatReg ? TYP_DOUBLE : TYP_LONG;
+                break;
+#endif // TARGET_64BIT
+            default:
+                assert("unexpected write to the stack.");
```

Comment:
This assert will never fire (will always evaluate to "true")

Reply:
Thanks, fixed!

Reply:
We have a few hits of this it seems:
```
src\coreclr\jit\valuenum.cpp
8811:                        noway_assert("LOCKADD should not appear before lowering");

src\coreclr\inc\clr_std\vector
296:                    assert("push_back: overflow");
314:                        assert("resize: overflow");
```

Reply:
> noway_assert

yes, I was fixing it in https://github.com/dotnet/runtime/issues/44132 but we don't have a check for it and add more from time to time.
I guess we can define a deleted "assert(const char*)" and it will give us a compilation failure, won't it?